blackboxopt/optimizers/testing.py

```diff
@@ -148,6 +150,81 @@ def is_deterministic_with_fixed_seed(
     return True
 
 
+def is_deterministic_when_reporting_shuffled_evaluations(
+    optimizer_class: Union[
+        Type[SingleObjectiveOptimizer], Type[MultiObjectiveOptimizer]
+    ],
+    optimizer_kwargs: dict,
+) -> bool:
+    """Check if determinism isn't affected by the order of initially reported data.
+
+    Repeatedly initialize the optimizer with the same parameter space and a fixed seed.
+    Report a set of initial evaluations in randomized order as initial data. Start
+    optimizing and check if the generated configurations for all optimizers are equal.
```

Comment:
Are we confident that reporting custom evaluations initially before optimization is the only relevant case we need to test for order invariance? In principle, the same situation can occur also at any other stage of optimization, consider e.g. parallel optimization runs where the runtime of the parallel evaluation jobs is not deterministic.

Reply:
>  In principle, the same situation can occur also at any other stage of optimization, consider e.g. parallel optimization runs where the runtime of the parallel evaluation jobs is not deterministic.

That's absolutely true, I'm very certain that the additional test is _not_ extensively covering all ordering related determinism issue, my motivation was only to extend coverage for the cases we observed so far.

Regarding the scenario of parallel optimization that you described, do you think there is a reasonable _reference_ test we could design? My understanding is, that the degree of determinism in such a case is very likely dependent on the type of optimizer (e.g. if the evaluations are processed staged/batched), is this correct?

I'm also not sure, to which degree we should _aim_ to be deterministic. That's something we might want to discuss in a bigger round, I guess.

Do you see any specific scenarios we _should_ cover here in _this_ PR?

Reply:
If this covers the issues we've seen that's good enough for me to go ahead with this PR. Especially since you're spot on about the challenges to find a representative, general enough reference test that would apply and discover issues in more individual optimization behavior.